src/Framework/Sdk/SdkResultItem.cs

```diff
@@ -34,35 +40,24 @@ public override bool Equals(object obj)
         {
             if (obj is SdkResultItem item &&
                    ItemSpec == item.ItemSpec &&
-                   Metadata?.Count == item.Metadata?.Count)
+                   item.Metadata is not null &&
+                   Metadata?.Count == item.Metadata.Count)
             {
-                if (Metadata != null)
-                {
-                    foreach (var kvp in Metadata)
-                    {
-                        if (item.Metadata[kvp.Key] != kvp.Value)
-                        {
-                            return false;
-                        }
-                    }
-                }
-
-                return true;
+                return Metadata.All(m => item.Metadata.TryGetValue(m.Key, out var itemValue) && itemValue == m.Value);
             }
             return false;
         }
 
         public override int GetHashCode()
         {
             int hashCode = -849885975;
-            hashCode = (hashCode * -1521134295) + EqualityComparer<string>.Default.GetHashCode(ItemSpec);
+            hashCode = hashCode ^ ItemSpec.GetHashCode();
 
             if (Metadata != null)
             {
                 foreach (var kvp in Metadata)
                 {
-                    hashCode = (hashCode * -1521134295) + kvp.Key.GetHashCode();
-                    hashCode = (hashCode * -1521134295) + kvp.Value.GetHashCode();
+                    hashCode ^= StringComparer.OrdinalIgnoreCase.GetHashCode($"{kvp.Key}: {kvp.Value ?? "V"}");
```

Comment:
I think this was the complete suggestion (to avoid the allocation)
```suggestion
                    hashCode ^= StringComparer.OrdinalIgnoreCase.GetHashCode(kvp.Key) * (StringComparer.OrdinalIgnoreCase.GetHashCode(kvp.Value ?? "V") + 1);
```